File: src/Butil/Bit.Butil/Scripts/streams.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    interface ReadableEntry {
        stream: ReadableStream;
        reader?: ReadableStreamDefaultReader;
        // The fetch behind a response body, so cancelling the stream stops the download instead of
        // just letting go of it.
        fetch?: AbortController;
        fetchCleanup?: () => void;
        // A pipe locks both of its ends for its whole duration, so nothing can be cancelled
        // through the stream itself while this is set - aborting it is the only way to stop it.
        pipe?: AbortController;
    }
    interface WritableEntry { stream: WritableStream; writer?: WritableStreamDefaultWriter; }

    const _readables: { [id: string]: ReadableEntry } = {};
    const _writables: { [id: string]: WritableEntry } = {};
    // transform id -> the ids its two ends are registered under, so releasing the transform
    // releases both.
    const _transforms: { [id: string]: { readableId: string; writableId: string } } = {};

    function trackReadable(id: string, stream: ReadableStream) { _readables[id] = { stream }; }
    function trackWritable(id: string, stream: WritableStream) { _writables[id] = { stream }; }

    butil.streams = {
        isSupported() { return typeof (window as any).ReadableStream === 'function'; },
        isTransformSupported() { return typeof (window as any).CompressionStream === 'function'; },

        // The one source: a fetch response body, read as it arrives instead of buffered whole.
        async fromResponse(id: string, url: string, req: any) {
            if (typeof (window as any).ReadableStream !== 'function') return null;

            // The request is built by the fetch module rather than here, so there is one mapping of
            // FetchRequest onto RequestInit - including how a shared AbortSignal composes with the
            // request's own controller, which is what cancel() reaches. Building it inside the try
            // along with the call is deliberate: an invalid header name or an unusable body throws
            // exactly where fetch() would fail, and a caller reading the documented Error result
            // should not have to catch a rejected promise for those two.
            const controller = new AbortController();
            let cleanup = () => { /* nothing was built yet */ };
            let response: Response;
            try {
                const built = butil.fetch.requestInit(req ?? {}, controller);
                cleanup = built.cleanup;
                response = await fetch(url, built.init);
            }
            catch (e: any) { cleanup(); return { ok: false, status: 0, error: e?.message ?? String(e) }; }

            // A 204, a HEAD, or an opaque no-cors response has no body at all - which is not an
            // error, but there is no stream to hand back either. Nothing gets registered, so the
            // listener the request put on a shared signal has to be let go of right here - there
            // will be no entry for cancel() or release() to do it through later.
            if (!response.body) {
                cleanup();
                return { ok: false, status: response.status, error: 'The response has no body.' };
            }

            trackReadable(id, response.body);
            _readables[id].fetch = controller;
            _readables[id].fetchCleanup = cleanup;
            const length = response.headers.get('content-length');

            return {
                ok: true,
                status: response.status,
                statusText: response.statusText,
                url: response.url,
                totalBytes: length ? Number(length) : null,
                error: null
            };
        },

        // Sink into .NET. write() returns the dispatch promise, so the stream will not pull the
        // next chunk until .NET has taken this one - which is what makes back-pressure reach across
        // the interop boundary rather than stopping at it.
        createWritable(dotNetRef: any, id: string, highWaterMark: number) {
            if (typeof (window as any).WritableStream !== 'function') return false;

            const stream = new WritableStream({
                write(chunk: any) {
                    const bytes = chunk instanceof Uint8Array ? chunk : new Uint8Array(chunk);
                    return butil.utils.dispatch(dotNetRef, 'InvokeSinkChunk', id, bytes);
                },
                close() { return butil.utils.dispatch(dotNetRef, 'InvokeSinkClose', id, null); },
                abort(reason: any) {
                    return butil.utils.dispatch(dotNetRef, 'InvokeSinkClose', id,
                        reason instanceof Error ? reason.message : String(reason ?? 'aborted'));
                }
            }, new CountQueuingStrategy({ highWaterMark: highWaterMark > 0 ? highWaterMark : 1 }));

            trackWritable(id, stream);
            return true;
        },

        createCompression(id: string, readableId: string, writableId: string, format: string, decompress: boolean) {
            const Ctor = (window as any)[decompress ? 'DecompressionStream' : 'CompressionStream'];
            if (typeof Ctor !== 'function') return false;

            let transform: TransformStream;
            try { transform = new Ctor(format); } catch { return false; } // unknown format

            trackReadable(readableId, transform.readable);
            trackWritable(writableId, transform.writable);
            _transforms[id] = { readableId, writableId };
            return true;
        },

        // Pull one chunk. The reader is acquired on first use, which is also what locks the stream -
        // tee() and pipeThrough() are unavailable from then on, by the specification's rules rather
        // than ours.
        async read(id: string) {
            const entry = _readables[id];
            if (!entry) return { done: true, data: null, error: 'unknown stream' };

            try {
                entry.reader = entry.reader ?? entry.stream.getReader();
                const { value, done } = await entry.reader.read();
                if (done) return { done: true, data: null, error: null };
                return { done: false, data: value instanceof Uint8Array ? value : new Uint8Array(value), error: null };
            } catch (e: any) {
                return { done: true, data: null, error: e?.message ?? String(e) };
            }
        },

        // Two streams from one, each getting every chunk. The original is locked afterwards and is
        // no longer readable itself - which is the point: it has been split, not copied.
        tee(id: string, firstId: string, secondId: string) {
            const entry = _readables[id];
            if (!entry || entry.reader) return false;   // already locked by a read()

            try {
                const [first, second] = entry.stream.tee();
                trackReadable(firstId, first);
                trackReadable(secondId, second);
                delete _readables[id];                  // the original is spent
                return true;
            } catch { return false; }
        },

        pipeThrough(id: string, transformId: string, resultId: string) {
            const entry = _readables[id];
            const transform = _transforms[transformId];
            if (!entry || entry.reader || !transform) return false;

            const writable = _writables[transform.writableId]?.stream;
            const readable = _readables[transform.readableId]?.stream;
            if (!writable || !readable) return false;

            try {
                const piped = entry.stream.pipeThrough({ writable, readable });
                // The piped stream *is* the transform's readable end, which createCompression
                // already registered under an id of its own. Two ids for one stream means reading
                // through the stale one locks the other, so the transform's entry goes away here.
                delete _readables[transform.readableId];
                trackReadable(resultId, piped);
                delete _readables[id];
                return true;
            } catch { return false; }
        },

        // Runs the whole pipe and settles when it is done. The promise is the completion signal:
        // everything a pipe does - back-pressure, closing the destination, propagating an error
        // from either end - happens inside it.
        async pipeTo(id: string, writableId: string, preventClose: boolean) {
            const entry = _readables[id];
            const destination = _writables[writableId];
            if (!entry || entry.reader || entry.pipe || !destination) return 'unknown or locked stream';

            // Without a signal a pipe cannot be stopped at all: it holds the lock on both ends, so
            // cancel() and abortWritable() are refused for as long as it runs, and an abandoned
            // download would keep going to completion.
            const pipe = new AbortController();
            entry.pipe = pipe;

            try {
                await entry.stream.pipeTo(destination.stream, { preventClose, signal: pipe.signal });
                delete _readables[id];
                return null;
            } catch (e: any) {
                return pipe.signal.aborted ? 'aborted' : (e?.message ?? String(e));
            } finally {
                entry.pipe = undefined;
            }
        },

        locked(id: string) {
            const entry = _readables[id];
            if (entry) return entry.stream.locked;
            return _writables[id]?.stream.locked ?? false;
        },

        async cancel(id: string, reason: string | null) {
            const entry = _readables[id];
            if (!entry) return;
            delete _readables[id];

            // Aborting the fetch is what actually stops a download; cancelling the body alone lets
            // the response keep arriving into nothing.
            try { entry.fetch?.abort(reason ?? undefined); } catch { /* already aborted */ }
            entry.fetchCleanup?.();

            try {
                // A pipe owns the lock on both ends, so its own signal is the only way in. A reader
                // has to be cancelled through itself; cancelling the stream under a live reader
                // throws for being locked.
                if (entry.pipe) entry.pipe.abort(reason ?? undefined);
                else if (entry.reader) await entry.reader.cancel(reason ?? undefined);
                else await entry.stream.cancel(reason ?? undefined);
            } catch { /* already closed or errored */ }
        },

        async write(id: string, bytes: Uint8Array) {
            const entry = _writables[id];
            if (!entry) return false;
            try {
                entry.writer = entry.writer ?? entry.stream.getWriter();
                // Awaiting ready before writing is what makes back-pressure real for a caller that
                // writes in a loop; without it every write is accepted and queued.
                await entry.writer.ready;
                await entry.writer.write(butil.utils.arrayToBuffer(bytes));
                return true;
            } catch { return false; }
        },

        async closeWritable(id: string) {
            const entry = _writables[id];
            if (!entry) return false;
            try {
                if (entry.writer) await entry.writer.close();
                else await entry.stream.close();
                delete _writables[id];
                return true;
            } catch { return false; }
        },

        async abortWritable(id: string, reason: string | null) {
            const entry = _writables[id];
            if (!entry) return false;
            try {
                if (entry.writer) await entry.writer.abort(reason ?? undefined);
                else await entry.stream.abort(reason ?? undefined);
                delete _writables[id];
                return true;
            } catch { return false; }
        },

        // Drops registry entries without cancelling: a stream handed to a pipe is no longer this
        // registry's business, and cancelling it here would break the pipe.
        release(id: string) {
            // Letting go of a request's stream without cancelling still means letting go of the
            // listener it put on a shared signal, which would otherwise hold this request for as
            // long as that signal lives.
            _readables[id]?.fetchCleanup?.();

            delete _readables[id];
            delete _writables[id];
            const transform = _transforms[id];
            if (transform) {
                delete _readables[transform.readableId];
                delete _writables[transform.writableId];
                delete _transforms[id];
            }
        },

        disposeAll() {
            for (const id of Object.keys(_readables)) butil.streams.cancel(id, 'disposed');
            for (const id of Object.keys(_writables)) butil.streams.abortWritable(id, 'disposed');
            for (const id of Object.keys(_transforms)) delete _transforms[id];
        }
    };
}(BitButil));

File: src/Butil/Bit.Butil/Publics/Streams.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/Streams_API">Streams API</see>:
/// data that arrives in pieces and is handled in pieces, rather than waited for and then held whole
/// in memory.
/// </summary>
/// <remarks>
/// The package already used streams internally - <see cref="Fetch"/> reads a response body through a
/// reader to report progress, and <see cref="Compression"/> pipes through the browser's codec. This
/// exposes them, so a pipeline can be yours: read a response as it arrives, split it in two, run it
/// through the native gzip codec, and hand the bytes to C# as they come.
/// <br/>
/// <b>The transforms are the browser's.</b> <see cref="CreateCompression"/> and
/// <see cref="CreateDecompression"/> wrap <c>CompressionStream</c>/<c>DecompressionStream</c>, and
/// those are what <see cref="ReadableStreamHandle.PipeThrough"/> accepts. C# code in the middle of a
/// pipeline would mean an interop hop per chunk; write it as read-transform-write around the pipe
/// instead, which costs the same and reads more honestly.
/// <br/>
/// Streams have an ownership rule that surprises people coming from .NET: reading from a stream
/// <em>locks</em> it, and a locked stream can no longer be teed or piped. That is the specification's
/// rule, not this wrapper's, and it is what keeps two consumers from silently stealing each other's
/// chunks.
/// </remarks>
[ButilService(typeof(Streams))]
public class Streams(IJSRuntime js) : IAsyncDisposable
{
    internal const string ChunkMethodName = nameof(InvokeSinkChunk);
    internal const string CloseMethodName = nameof(InvokeSinkClose);

    private readonly ConcurrentDictionary<Guid, SinkHandlers> _sinks = new();

    // Per-instance callback reference (see Keyboard): sinks are isolated per circuit / WASM app and
    // released on disposal - no static state, no cross-circuit leak.
    private DotNetObjectReference<Streams>? _dotNetRef;
    private DotNetObjectReference<Streams> DotNetRef => DotNetObjectReferenceHelper.GetOrCreate(ref _dotNetRef, this);

    private sealed record SinkHandlers(Action<byte[]> OnChunk, Action<string?>? OnFinished);

    /// <summary>True when the runtime exposes <c>ReadableStream</c>.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (<c>false</c>) rather than
    /// throwing, so the result can't be distinguished from a genuine value. If you branch on it,
    /// defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.streams.isSupported");

    /// <summary>
    /// True when the runtime exposes <c>CompressionStream</c> - which is what
    /// <see cref="CreateCompression"/> and <see cref="CreateDecompression"/> need.
    /// </summary>
    public ValueTask<bool> IsTransformSupported() => js.Invoke<bool>("BitButil.streams.isTransformSupported");

    /// <summary>
    /// Invoked from JS for each chunk written into a sink. Public + <see cref="JSInvokableAttribute"/>
    /// so it can be dispatched through the per-instance <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    /// <remarks>
    /// The stream waits for this call to return before pulling the next chunk, which is how
    /// back-pressure reaches across the interop boundary instead of stopping at it. A slow handler
    /// slows the producer, as it should.
    /// </remarks>
    [JSInvokable(ChunkMethodName)]
    public void InvokeSinkChunk(Guid id, byte[] chunk)
    {
        if (_sinks.TryGetValue(id, out var handlers)) handlers.OnChunk(chunk);
    }

    /// <summary>
    /// Invoked from JS when a sink closes or aborts. Public + <see cref="JSInvokableAttribute"/> so
    /// it can be dispatched through the per-instance <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    [JSInvokable(CloseMethodName)]
    public void InvokeSinkClose(Guid id, string? reason)
    {
        // The sink is finished either way - drop it here rather than waiting for a Dispose the
        // caller has no reason to make once the stream has ended.
        if (_sinks.TryRemove(id, out var handlers)) handlers.OnFinished?.Invoke(reason);
    }

    /// <summary>
    /// Starts a request and hands back its body as a stream, before the body has arrived.
    /// </summary>
    /// <param name="request">
    /// The same shape <see cref="Fetch.Send"/> takes, including
    /// <see cref="FetchRequest.Signal"/> - a shared abort signal cancels the download as it would
    /// cancel any other request.
    /// </param>
    /// <returns>
    /// The response's status and headers, with <see cref="StreamedResponse.Stream"/> set when there
    /// is a body to read. A failed request or a bodyless response (a 204, a <c>HEAD</c>, an opaque
    /// <c>no-cors</c> response) comes back with <see cref="StreamedResponse.Error"/> set rather than
    /// throwing - a request that does not arrive is a normal outcome.
    /// </returns>
    /// <remarks>
    /// This is the difference between "download it, then look at it" and "look at it as it
    /// downloads": a gigabyte through <see cref="Fetch.Send"/> is a gigabyte of managed memory,
    /// while a gigabyte through here is one chunk at a time.
    /// </remarks>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(FetchRequest))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(StreamedResponseDto))]
    public async ValueTask<StreamedResponse> FromResponse(FetchRequest request)
    {
        ArgumentNullException.ThrowIfNull(request);

        var id = Guid.NewGuid();
        var result = await js.Invoke<StreamedResponseDto?>("BitButil.streams.fromResponse", id, request.Url, request);

        if (result is null)
            return new StreamedResponse(null, 0, string.Empty, request.Url, null, "The runtime has no Streams API.");

        return new StreamedResponse(
            result.Ok ? new ReadableStreamHandle(js, id) : null,
            result.Status,
            result.StatusText ?? string.Empty,
            result.Url ?? request.Url,
            result.TotalBytes,
            result.Error);
    }

    /// <summary>
    /// Creates a stream that writes into C#: everything piped to it arrives at
    /// <paramref name="onChunk"/>.
    /// </summary>
    /// <param name="onChunk">
    /// Called for each chunk, in order. The producer waits for this to return before sending the
    /// next one, so a slow handler slows the download rather than queueing behind it.
    /// </param>
    /// <param name="onFinished">
    /// Called once when the stream ends: with null after a clean close, or with the reason when it
    /// was aborted.
    /// </param>
    /// <param name="highWaterMark">How many chunks may be queued before the producer is asked to wait. One is the honest default for a sink whose consumer is on the other side of an interop call.</param>
    /// <returns>A handle, or null when the runtime has no <c>WritableStream</c>.</returns>
    public async ValueTask<WritableStreamHandle?> CreateWritable(Action<byte[]> onChunk,
                                                                 Action<string?>? onFinished = null,
                                                                 int highWaterMark = 1)
    {
        ArgumentNullException.ThrowIfNull(onChunk);

        var id = Guid.NewGuid();
        _sinks[id] = new SinkHandlers(onChunk, onFinished);

        bool created;
        try
        {
            created = await js.Invoke<bool>("BitButil.streams.createWritable", DotNetRef, id, highWaterMark);
        }
        catch
        {
            // The sink was registered before the call, so a call that throws has to take it back out
            // again - nothing else will ever reach handlers for a stream that was never created.
            _sinks.TryRemove(id, out _);
            throw;
        }

        if (created is false)
        {
            _sinks.TryRemove(id, out _);
            return null;
        }

        return new WritableStreamHandle(js, id, () => _sinks.TryRemove(id, out _));
    }

    /// <summary>
    /// A transform that compresses everything passing through it, using the browser's own codec.
    /// </summary>
    /// <param name="format">Which codec. Defaults to gzip, the interoperable one.</param>
    /// <returns>A handle, or null when the runtime has no <c>CompressionStream</c>.</returns>
    /// <remarks>
    /// Unlike <see cref="Compression.Compress"/>, which takes bytes and returns bytes, this one
    /// never holds the whole payload: it is a stage in a pipeline.
    /// </remarks>
    public ValueTask<TransformStreamHandle?> CreateCompression(CompressionFormat format = CompressionFormat.Gzip)
        => CreateTransform(format, decompress: false);

    /// <summary>
    /// A transform that decompresses everything passing through it.
    /// </summary>
    /// <param name="format">The codec the data was compressed with. Defaults to gzip.</param>
    /// <returns>A handle, or null when the runtime has no <c>DecompressionStream</c>.</returns>
    /// <remarks>
    /// A format that does not match the data errors the pipe rather than this call - corrupt input
    /// is only discovered while reading it.
    /// </remarks>
    public ValueTask<TransformStreamHandle?> CreateDecompression(CompressionFormat format = CompressionFormat.Gzip)
        => CreateTransform(format, decompress: true);

    private async ValueTask<TransformStreamHandle?> CreateTransform(CompressionFormat format, bool decompress)
    {
        var id = Guid.NewGuid();
        var readableId = Guid.NewGuid();
        var writableId = Guid.NewGuid();

        var created = await js.Invoke<bool>("BitButil.streams.createCompression",
            id, readableId, writableId, ToName(format), decompress);

        return created ? new TransformStreamHandle(js, id, readableId, writableId) : null;
    }

    // The strings CompressionStream's constructor accepts - the same mapping Compression uses.
    private static string ToName(CompressionFormat format) => format switch
    {
        CompressionFormat.Deflate => "deflate",
        CompressionFormat.DeflateRaw => "deflate-raw",
        _ => "gzip",
    };

    /// <summary>
    /// On scope/circuit teardown, cancels every stream whose handle was never disposed, so an
    /// abandoned download stops rather than running to completion into nothing.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            _sinks.Clear();
            await js.InvokeVoid("BitButil.streams.disposeAll");
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed
        finally
        {
            _dotNetRef?.Dispose();
            _dotNetRef = null;
        }
        GC.SuppressFinalize(this);
    }
}
